### ISSUE TITLE
Horovod + Apex with backward_passes_per_step

### ISSUE THREAD

**Issue by agemagician** [horovod/horovod#1535]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
Pytorch
2. Framework version:
Pytorch 1.1.0
3. Horovod version:
Latest from Github
4. MPI version:
IBM MPI Spectrum
5. CUDA version:
10
6. NCCL version:
2
7. Python version:
3.6
8. OS and version:
IBM Power
9. GCC version:
Unkown

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
yes
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
yes
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?
yes

**Your question:**
I am running a pytorch code for training a model and I am using Horovod with apex.
I want to use gradient accumulation with Horovod.

Q1:
Is backward_passes_per_step is exactly equivalent to gradient accumulation ?

Q2:
Giving that I had to change the normal structure of Horovod training to make it work with Nvidia apex as recommended in:
https://gist.github.com/alsrgv/0713add50fe49a409316832a31612dde
as follows:
```
# Horovod: wrap optimizer with DistributedOptimizer.
optimizer = hvd.DistributedOptimizer(optimizer,
                                     named_parameters=model.named_parameters(),
                                     compression=compression,
                                     backward_passes_per_step=args.batches_per_allreduce)

# Horovod: broadcast parameters & optimizer state.
hvd.broadcast_parameters(model.state_dict(), root_rank=0)
hvd.broadcast_optimizer_state(optimizer, root_rank=0)

# Apex
model, optimizer = amp.initialize(model, optimizer, opt_level="O1")
```
```
# Apex
    acc_loss = loss/args.batches_per_allreduce
    with amp.scale_loss(acc_loss, optimizer) as scaled_loss:
        scaled_loss.backward()
        optimizer.synchronize()
    torch.nn.utils.clip_grad_norm_(amp.master_params(optimizer), args.clip)
    if batch_num%args.batches_per_allreduce == 0:
            with optimizer.skip_synchronize():
                optimizer.step()
                model.zero_grad()
```
Is the backward_passes_per_step going to work or since I call synchronize during loss. It will force Horovod to not use backward_passes_per_step ?

**Comment by tgaddair** [comment]
Hey @agemagician, to answer your first question, `backward_passes_per_step` is the parameter for local gradient accumulation, yes.

Regarding Apex, @alsrgv or @romerojosh may be able to give you more information.

**Comment by jhsa26** [comment]
> Hey @agemagician, to answer your first question, `backward_passes_per_step` is the parameter for local gradient accumulation, yes.
> 
> Regarding Apex, @alsrgv or @romerojosh may be able to give you more information.

if back_ward_passes_per_step is larger than local gradient accumulation, the result is the same as that of back_ward_passes_per_step equal to local gradient accumulation. I guess that when back_ward_passes_per_step is equal to local gradient accumulation, the computed local gradients
are exchanged with other ranks while computing gradients of preceding  layer.
